Crash when reopening the app with certain dialogs in the browser and >60,000 cards selected

### Checked for duplicates?

- [x] This issue is not a duplicate

### Does it also happen in the desktop version?

- [x] This bug does not occur in the latest version of Anki Desktop

### What are the steps to reproduce this bug?

1. Make a deck with 60,000 or more cards, 70,000 to be sure.
2. Open the browser and select all.
3. Open either the set due date or the export dialogs.
4. Leave the app, but don't remove it from recents.
5. Reopen the app.

You'll either be sent to the main screen or the app won't open and the bug report toast will appear.

### Expected behaviour

No crash.

### Debug info

```text
AnkiDroid Version = 2.24.0alpha9 (72568b4e67dfb2f89f8bb8d0923dd6460d3e6e49)  
Backend Version = 0.1.63-anki25.09.2 (25.09.2 3890e12c9e48c028c3f12aa58cb64bd9f8895e30)  
Android Version = 13 (SDK 33)  
ProductFlavor = full  
Device Info = Xiaomi | POCO | mondrian | mondrian_global | 23013PC75G | qcom  
WebView Info = [com.google.android.webview | 763215903]: Mozilla/5.0 (Linux; Android 13; 23013PC75G Build/TKQ1.220905.001; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/145.0.7632.159 Mobile Safari/537.36  
ACRA UUID = 4afc3b3f-e93f-4a66-bc71-2b9476ef9eb0  
FSRS = 5.1.0 (Enabled: true)  
Crash Reports Enabled = true
```

### (Optional) Anything else you want to share?

I reproduced it in v2.22.3 too.

### Research

- [x] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [x] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

These are the **LogCat**:

```05-03 23:06:41.764 20326 20326 E AndroidRuntime: FATAL EXCEPTION: main
05-03 23:06:41.764 20326 20326 E AndroidRuntime: Process: com.ichi2.anki.debug, PID: 20326
05-03 23:06:41.764 20326 20326 E AndroidRuntime: java.lang.RuntimeException: android.os.TransactionTooLargeException: data parcel size 619604 bytes
05-03 23:06:41.764 20326 20326 E AndroidRuntime: Bundle stats:
05-03 23:06:41.764 20326 20326 E AndroidRuntime:   android:viewHierarchyState [size=2440]
05-03 23:06:41.764 20326 20326 E AndroidRuntime:     android:views [size=2392]
05-03 23:06:41.764 20326 20326 E AndroidRuntime:   androidx.lifecycle.BundlableSavedStateRegistry.key [size=616368]
05-03 23:06:41.764 20326 20326 E AndroidRuntime:     androidx.lifecycle.internal.SavedStateHandlesProvider [size=1216]
05-03 23:06:41.764 20326 20326 E AndroidRuntime:     android:support:activity-result [size=1900]
05-03 23:06:41.764 20326 20326 E AndroidRuntime:       KEY_COMPONENT_ACTIVITY_REGISTERED_KEYS [size=1416]
05-03 23:06:41.764 20326 20326 E AndroidRuntime:     android:support:fragments [size=612884]
05-03 23:06:41.764 20326 20326 E AndroidRuntime:       fragment_ad2350b8-60c2-4f25-996a-90ffd66c7b2d [size=610624]
05-03 23:06:41.764 20326 20326 E AndroidRuntime:         arguments [size=600964]
05-03 23:06:41.764 20326 20326 E AndroidRuntime:           arg_exported_ids [size=600648]
05-03 23:06:41.764 20326 20326 E AndroidRuntime:         savedInstanceState [size=8932]
05-03 23:06:41.764 20326 20326 E AndroidRuntime:           android:savedDialogState [size=8860]
05-03 23:06:41.764 20326 20326 E AndroidRuntime:             android:dialogHierarchy [size=8736]
05-03 23:06:41.764 20326 20326 E AndroidRuntime:               android:views [size=8688]
05-03 23:06:41.764 20326 20326 E AndroidRuntime:       fragment_d9d5afa7-5b88-4e1c-86bf-8ba97834529b [size=1584]
05-03 23:06:41.764 20326 20326 E AndroidRuntime: PersistableBundle stats:
05-03 23:06:41.764 20326 20326 E AndroidRuntime:   [null]
05-03 23:06:41.764 20326 20326 E AndroidRuntime:        at android.app.servertransaction.PendingTransactionActions$StopInfo.run(PendingTransactionActions.java:146)
05-03 23:06:41.764 20326 20326 E AndroidRuntime:        at android.os.Handler.handleCallback(Handler.java:958)
05-03 23:06:41.764 20326 20326 E AndroidRuntime:        at android.os.Handler.dispatchMessage(Handler.java:99)
05-03 23:06:41.764 20326 20326 E AndroidRuntime:        at android.os.Looper.loopOnce(Looper.java:222)
05-03 23:06:41.764 20326 20326 E AndroidRuntime:        at android.os.Looper.loop(Looper.java:314)
05-03 23:06:41.764 20326 20326 E AndroidRuntime:        at android.app.ActivityThread.main(ActivityThread.java:8788)
05-03 23:06:41.764 20326 20326 E AndroidRuntime:        at java.lang.reflect.Method.invoke(Native Method)
05-03 23:06:41.764 20326 20326 E AndroidRuntime:        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:569)
05-03 23:06:41.764 20326 20326 E AndroidRuntime:        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1090)
05-03 23:06:41.764 20326 20326 E AndroidRuntime: Caused by: android.os.TransactionTooLargeException: data parcel size 619604 bytes
05-03 23:06:41.764 20326 20326 E AndroidRuntime:        at android.os.BinderProxy.transactNative(Native Method)
05-03 23:06:41.764 20326 20326 E AndroidRuntime:        at android.os.BinderProxy.transact(BinderProxy.java:639)
05-03 23:06:41.764 20326 20326 E AndroidRuntime:        at android.app.IActivityClientController$Stub$Proxy.activityStopped(IActivityClientController.java:1459)
05-03 23:06:41.764 20326 20326 E AndroidRuntime:        at android.app.ActivityClient.activityStopped(ActivityClient.java:112)
05-03 23:06:41.764 20326 20326 E AndroidRuntime:        at android.app.servertransaction.PendingTransactionActions$StopInfo.run(PendingTransactionActions.java:135)```

--- a/AnkiDroid/src/main/java/com/ichi2/anki/browser/CardBrowserFragment.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/browser/CardBrowserFragment.kt
@@ -1335,7 +1335,12 @@ class CardBrowserFragment :
 
     fun exportSelected() {
         val (type, selectedIds) = activityViewModel.querySelectionExportData() ?: return
-        ExportDialogFragment.newInstance(type, selectedIds).show(parentFragmentManager, "exportDialog")
+        ExportDialogFragment
+            .newInstance(
+                requireContext().externalCacheDir ?: requireContext().cacheDir,
+                type,
+                selectedIds,
+            ).show(parentFragmentManager, "exportDialog")
     }
 
     fun showOptionsDialog() {

--- a/AnkiDroid/src/main/java/com/ichi2/anki/browser/CardBrowserViewModel.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/browser/CardBrowserViewModel.kt
@@ -1600,6 +1600,21 @@ class IdsFile(
     }
 }
 
+/** Attempt to delete the associated [IdsFile] and logs the result */
+fun IdsFile.removeSafely(owner: String) {
+    runCatching { delete() }
+        .onFailure { throwable ->
+            Timber.w(
+                throwable,
+                "Exception when removing IdsFile of $owner",
+            )
+        }.onSuccess { status ->
+            Timber.i(
+                "$owner associated IdsFile was deleted: $status",
+            )
+        }
+}
+
 /**
  * Determines if a card can be repositioned.
  *

--- a/AnkiDroid/src/main/java/com/ichi2/anki/browser/FindAndReplaceDialogFragment.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/browser/FindAndReplaceDialogFragment.kt
@@ -178,15 +178,7 @@ class FindAndReplaceDialogFragment : AnalyticsDialogFragment() {
 
     /** Attempt to delete the associated [IdsFile] and logs the result */
     private fun removeIdsFile() {
-        runCatching { idsFile.delete() }
-            .onFailure { throwable ->
-                Timber.w(
-                    throwable,
-                    "Exception when removing IdsFile of FindAndReplaceDialogFragment",
-                )
-            }.onSuccess { status ->
-                Timber.i("FindAndReplaceDialogFragment associated IdsFile was deleted: $status")
-            }
+        idsFile.removeSafely(TAG)
     }
 
     companion object {

--- a/AnkiDroid/src/main/java/com/ichi2/anki/export/ExportDialogFragment.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/export/ExportDialogFragment.kt
@@ -15,6 +15,7 @@ package com.ichi2.anki.export
 
 import android.app.Dialog
 import android.content.Context
+import android.content.DialogInterface
 import android.os.Bundle
 import android.view.View
 import android.view.ViewGroup
@@ -38,6 +39,8 @@ import com.ichi2.anki.CollectionManager
 import com.ichi2.anki.CollectionManager.TR
 import com.ichi2.anki.CollectionManager.withCol
 import com.ichi2.anki.R
+import com.ichi2.anki.browser.IdsFile
+import com.ichi2.anki.browser.removeSafely
 import com.ichi2.anki.common.time.TimeManager
 import com.ichi2.anki.common.time.getTimestamp
 import com.ichi2.anki.compat.CompatHelper.Companion.getSerializableCompat
@@ -50,6 +53,7 @@ import com.ichi2.anki.libanki.DeckId
 import com.ichi2.anki.libanki.DeckNameId
 import com.ichi2.anki.requireAnkiActivity
 import com.ichi2.anki.ui.BasicItemSelectedListener
+import com.ichi2.anki.utils.ext.requireParcelable
 import com.ichi2.utils.negativeButton
 import com.ichi2.utils.positiveButton
 import kotlinx.coroutines.launch
@@ -61,6 +65,13 @@ import java.io.File
  */
 class ExportDialogFragment : DialogFragment() {
     private lateinit var binding: DialogExportOptionsBinding
+
+    override fun onDismiss(dialog: DialogInterface) {
+        super.onDismiss(dialog)
+        if (arguments?.containsKey(ARG_IDS_FILE) == true) {
+            removeIdsFile()
+        }
+    }
 
     override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
         binding = DialogExportOptionsBinding.inflate(requireActivity().layoutInflater, null, false)
@@ -327,17 +338,15 @@ class ExportDialogFragment : DialogFragment() {
     private fun buildExportLimit(): ExportLimit =
         when (arguments?.getSerializableCompat<ExportType>(ARG_TYPE)) {
             ExportType.Notes -> {
-                val selectedNotesIds =
-                    arguments?.getLongArray(ARG_EXPORTED_IDS)
-                        ?: error("Requested export for selected notes but no notes ids were passed in!")
-                exportLimit { noteIds = noteIds { this.noteIds.addAll(selectedNotesIds.toList()) } }
+                val ids = requireArguments().requireParcelable<IdsFile>(ARG_IDS_FILE).getIds()
+
+                exportLimit { noteIds = noteIds { this.noteIds.addAll(ids) } }
             }
 
             ExportType.Cards -> {
-                val selectedCardIds =
-                    arguments?.getLongArray(ARG_EXPORTED_IDS)
-                        ?: error("Requested export for selected cards but no cards ids were passed in!")
-                exportLimit { cardIds = cardIds { this.cids.addAll(selectedCardIds.toList()) } }
+                val ids = requireArguments().requireParcelable<IdsFile>(ARG_IDS_FILE).getIds()
+
+                exportLimit { cardIds = cardIds { this.cids.addAll(ids) } }
             }
             // notes/cards weren't selected so export the chosen decks
             null -> {
@@ -351,6 +360,13 @@ class ExportDialogFragment : DialogFragment() {
                 }
             }
         }
+
+    /** Attempt to delete the associated [IdsFile] and logs the result */
+    private fun removeIdsFile() {
+        val idsFile = requireArguments().requireParcelable<IdsFile>(ARG_IDS_FILE)
+
+        idsFile.removeSafely("ExportDialogFragment")
+    }
 
     private fun getExportRootFile() =
         File(requireActivity().externalCacheDir, "export").also {
@@ -420,7 +436,7 @@ class ExportDialogFragment : DialogFragment() {
     companion object {
         private const val ARG_DECK_ID = "arg_deck_id"
         private const val ARG_TYPE = "arg_type"
-        private const val ARG_EXPORTED_IDS = "arg_exported_ids"
+        private const val ARG_IDS_FILE = "arg_ids_file"
 
         /**
          * Create a new instance of this dialog without any initial constraints(for example when
@@ -440,13 +456,16 @@ class ExportDialogFragment : DialogFragment() {
          * Create a new instance of this dialog targeting a selection of cards or notes for export.
          */
         fun newInstance(
+            cacheDir: File,
             type: ExportType,
             ids: List<Long>,
         ) = ExportDialogFragment().apply {
+            val idsFile = IdsFile(cacheDir, ids, "export")
+
             arguments =
                 Bundle().apply {
                     putSerializable(ARG_TYPE, type)
-                    putLongArray(ARG_EXPORTED_IDS, ids.toLongArray())
+                    putParcelable(ARG_IDS_FILE, idsFile)
                 }
         }
     }

